scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala

```diff
@@ -296,6 +296,29 @@ object TreeSyntax {
       }
       !isOnlyChildOfOnlyChild(t)
     }
+    @annotation.tailrec
+    def guessNeedsLineSep(t: Tree): Boolean = t match {
+      case _: Decl.Val | _: Decl.Var | _: Decl.Def | _: Defn.Macro | _: Defn.Type | _: Term.This |
+          _: Term.Name | _: Term.Select | _: Member.Apply | _: Term.Ascribe | _: Term.Tuple |
+          _: Term.New | _: Term.Interpolate | _: Term.Xml | _: Lit =>
+        true
+      case t: Term.Do => false
+      case t: Tree.WithBody => guessNeedsLineSep(t.body)
```

Comment:
please remove `Defn.Macro` and `Defn.Type` above, they are already covered as they are both `Tree.WithBody`.

Reply:
The `Defn.Macro` was indeed redundant now.
The `Defn.Type` has a `body: Type`, so I can only remove it from here if I replace it with a new pattern `_: Type` instead (that should only be called from a recursive call of `Defn.Type` bodies).

Reply:
if `Defn.Type` (or `Defn.Macro`) should always be excluded (regardless of its body), then please feel free to keep it here. it looks like your tests are quite comprehensive, that should become clear right away.